examples/vision/simsiam.py

```diff
@@ -0,0 +1,434 @@
+target dataset with 5% or 10% labels present. Practically, we can use the backbone model
+for any downstream task such as semantic segmentation, object detection, and so on where
+the backbone models are usually pre-trained with *pure supervised learning*. 
+"""
+
+# We first create labeled `Dataset` objects.
+train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+
+# Then we shuffle, batch, and prefetch this dataset for performance. We
+# also apply random resized crops as an augmentation but only to the
+# training set.
+train_ds = (
+    train_ds.shuffle(1024)
+    .map(lambda x, y: (flip_random_crop(x), y), num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+test_ds = test_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+# Extract the backbone ResNet20.
+backbone = tf.keras.Model(
+    simsiam.encoder.input, simsiam.encoder.get_layer("backbone_pool").output
+)
+
+# We then create our linear classifier and train it.
+backbone.trainable = False
+inputs = layers.Input((CROP_TO, CROP_TO, 3))
+x = backbone(inputs, training=False)
+outputs = layers.Dense(10, activation="softmax")(x)
+linear_model = tf.keras.Model(inputs, outputs, name="linear_model")
+
+# Start training.
+linear_model.compile(
+    loss="sparse_categorical_crossentropy",
+    metrics=["accuracy"],
+    optimizer=tf.keras.optimizers.SGD(lr_decayed_fn, momentum=0.9),
+)
+history = linear_model.fit(
+    train_ds, validation_data=test_ds, epochs=EPOCHS, callbacks=[early_stopping]
+)
+_, test_acc = linear_model.evaluate(test_ds)
+print("Test accuracy: {:.2f}%".format(test_acc * 100))
+
+"""
+## Notes
+* More data and longer pre-training schedule benefit SSL in general.
+* SSL is particularly very helpful when you do not have access to very limited *labeled*
+training data but you can manage to build a large corpus of unlabeled data. Recently,
+using an SSL method called [SwAV](https://arxiv.org/abs/2006.09882), a group of
+researchers at Facebook trained a [RegNet](https://arxiv.org/abs/2006.09882) on 2 Billion
+images. They were able to achieve downstream performance very close to those achieved by
+pure supervised pre-training. For some downstream tasks, their method even outperformed
+the supervised counterparts. You can check out [their
+paper](https://arxiv.org/pdf/2103.01988.pdf) to know the details. 
+* If you are interested to understand why contrastive SSL helps networks learn meaningful
+representations, you can check out the following resources:
+   * [Self-supervised learning: The dark matter of
+intelligence](https://ai.facebook.com/blog/self-supervised-learning-the-dark-matter-of-intelligence/)
+   * [Understanding self-supervised learning using controlled datasets with known
+structure](https://sslneuips20.github.io/files/CameraReadys%203-77/64/CameraReady/Understanding_self_supervised_learning.pdf)
```

Comment:
Suggestion: some refactoring

```suggestion
## Notes

* More data and longer pre-training schedule can benefit SSL in general.
* SSL is particularly helpful when you have very limited *labeled* training
data but you can have a large corpus of unlabeled data. In an SSL method called
[SwAV](https://arxiv.org/abs/2006.09882), researchers trained a
[RegNet](https://arxiv.org/abs/2006.09882) on 2 billion images—their results in
downstream performance was very close to the one achieved by pure supervised
pre-training. In addition, for certain downstream tasks, their method managed to
outperform the supervised counterparts. (For more details, check 
[Goyal et al., 2021](https://arxiv.org/pdf/2103.01988.pdf).

To learn more about contrastive SSL and how it can help neural networks learn
meaningful representations:

* [Self-supervised learning: The dark matter of
intelligence](https://ai.facebook.com/blog/self-supervised-learning-the-dark-matter-of-intelligence/).
* [Understanding self-supervised learning using controlled datasets with known
structure](https://sslneuips20.github.io/files/CameraReadys%203-77/64/CameraReady/Understanding_self_supervised_learning.pdf).
```

@sayakpaul LMK if I got the gist right here!